cmd/libsnap-confine-private/cgroup-support.c

```diff
@@ -97,3 +98,123 @@ bool sc_cgroup_is_v2() {
     }
     return false;
 }
+
+static bool traverse_looking_for_prefix_in_dir(DIR *root, const char *prefix, const char *skip) {
+    while (true) {
+        errno = 0;
+        struct dirent *ent = readdir(root);
+        if (ent == NULL) {
+            if (errno != 0) {
+                die("cannot read directory entry");
+            }
+            break;
+        }
+        if (ent->d_type != DT_DIR) {
+            continue;
+        }
+        if (sc_streq(ent->d_name, "..") || sc_streq(ent->d_name, ".")) {
+            /* we don't want to go up or process the current directory again */
+            continue;
+        }
+        if (sc_streq(ent->d_name, skip)) {
+            continue;
+        }
+        if (sc_startswith(ent->d_name, prefix)) {
+            debug("found matching prefix in \"%s\"", ent->d_name);
+            /* the directory starts with our prefix */
+            return true;
+        }
+        int entfd = openat(dirfd(root), ent->d_name, O_DIRECTORY | O_NOFOLLOW | O_CLOEXEC);
+        if (entfd == -1) {
+            die("cannot open directory entry \"%s\"", ent->d_name);
+        }
+        debug("got directory fd: %d", entfd);
+        /* takes ownership o the file descriptor? */
+        DIR *entdir SC_CLEANUP(sc_cleanup_closedir) = fdopendir(entfd);
+        if (entdir == NULL) {
+            die("cannot fdopendir directory \"%s\"", ent->d_name);
+        }
+        debug("descend into %s", ent->d_name);
+        int found = traverse_looking_for_prefix_in_dir(entdir, prefix, skip);
+        if (found == true) {
+            return true;
+        }
+    }
+    return false;
+}
+
+bool sc_cgroup_v2_is_tracking_snap(const char *snap_instance) {
+    debug("is cgroup tracking snap %s?", snap_instance);
+    char tracking_group_name[PATH_MAX] = {0};
+    sc_must_snprintf(tracking_group_name, sizeof tracking_group_name, "snap.%s.", snap_instance);
+
+    char *own_group SC_CLEANUP(sc_cleanup_string) = sc_cgroup_v2_own_path_full();
+    if (own_group == NULL) {
+        die("cannot obtain own group path");
+    }
+    debug("own group: %s", own_group);
```

Comment:
(nitpick^9) I was thinking a bit about how our debug output looks like with all the new debug output. But I just ran a simple `test-snapd-tools.echo`  test and it's already ~100 lines so probably okay to have some extra lines in this already long output.

Reply:
I dropped some of the debug lines, as the output was indeed to noisy